hazelcast/src/main/java/com/hazelcast/config/OnJoinCacheConfig.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.spi.annotation.PrivateApi;
+
+import javax.cache.configuration.CacheEntryListenerConfiguration;
+import java.io.IOException;
+
+/**
+ * This subclass of {@link CacheConfig} is used to communicate cache configurations in pre-join cache operations when cluster
+ * version is at least 3.9. The key difference against {@link CacheConfig} is that the key/value class names are used in its
+ * serialized form, instead of the actual {@code Class} objects. Thus the actual key-value classes are only resolved when first
+ * used (by means of {@link CacheConfig#getKeyType()} or {@link CacheConfig#getValueType()}). This allows resolution of
+ * these classes from remote user code deployment repositories (which are not available while the pre-join operation is being
+ * deserialized and executed).
+ *
+ * @param <K> the key type of this cache configuration
+ * @param <V> the value type
+ * @since 3.9
+ */
+@PrivateApi
+public class OnJoinCacheConfig<K, V> extends CacheConfig<K, V> implements IdentifiedDataSerializable {
```

Comment:
Can't we keep `OnJoinCacheConfig ` in internal cache package, instead of `com.hazelcast.config`?

Reply:
Moved to `com.hazelcast.cache.impl` package.